streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -822,8 +828,51 @@ private void addStreamTasks(Collection<TopicPartition> assignment) {
             }
         }
 
-        // finally destroy any remaining suspended tasks
+        // destroy any remaining suspended tasks
         removeSuspendedTasks();
+
+        // create all newly assigned tasks (guard against race condition with other thread via backoff and retry)
+        // -> other thread will call removeSuspendedTasks(); eventually
+        long backoffTimeMs = 50L;
+        while (true) {
+            final Iterator<Map.Entry<TaskId,Set<TopicPartition>>> it = newTasks.entrySet().iterator();
+            while (it.hasNext()) {
+                final Map.Entry<TaskId, Set<TopicPartition>> newTaskAndPartitions = it.next();
+                final TaskId taskId = newTaskAndPartitions.getKey();
+                final Set<TopicPartition> partitions = newTaskAndPartitions.getValue();
+
+                try {
+                    log.debug("{} creating new task {}", logPrefix, taskId);
+                    final StreamTask task = createStreamTask(taskId, partitions);
+
+                    activeTasks.put(taskId, task);
+
+                    for (TopicPartition partition : partitions) {
+                        activeTasksByPartition.put(partition, task);
+                    }
+
+                    it.remove();
+                } catch (final ProcessorStateException e) {
```

Comment:
Just a thought, but i wonder if it would be better if we change the exception thrown from the failure to get the lock to be specific to that? So we know that it is only thrown in that case? I'm not sure if there is currently anywhere else in this flow that will throw an IOE wrapped in a PSE, but if we had a specific exception it would protect us going forward.